community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/CachingExpandInto.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_0.pipes
+
+import java.util.concurrent.ThreadLocalRandom
+
+import org.neo4j.cypher.internal.compiler.v3_0.ExecutionContext
+import org.neo4j.cypher.internal.compiler.v3_0.spi.QueryContext
+import org.neo4j.cypher.internal.frontend.v3_0.{InternalException, SemanticDirection}
+import org.neo4j.graphdb.{Relationship, Node}
+import org.neo4j.helpers.collection.PrefetchingIterator
+
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+/**
+ * Used by pipes that needs to expand between two known nodes.
+ *
+ * Given a pattern (a)-->(b) it will do the following:
+ * - if both a and b are dense nodes, start from the one with the lesser degree
+ * - if just one of the nodes is dense, start from the non-dense node
+ * - if both are non-dense, randomly pick one or the other
+ * - cache all found connecting relationships.
+ *
+ */
+trait CachingExpandInto {
+
+  /**
+   * Finds all relationships connecting fromNode and toNode.
+   */
+  protected def findRelationships(query: QueryContext, fromNode: Node, toNode: Node,
+                                relCache: RelationshipsCache, dir: SemanticDirection, relTypes: => Option[Seq[Int]]): Iterator[Relationship] = {
+
+    val fromNodeIsDense = query.nodeIsDense(fromNode.getId)
+    val toNodeIsDense = query.nodeIsDense(toNode.getId)
+
+    //if both nodes are dense, start from the one with the lesser degree
+    if (fromNodeIsDense && toNodeIsDense) {
+      //check degree and iterate from the node with smaller degree
+      val fromDegree = getDegree(fromNode, relTypes, dir, query)
+      if (fromDegree == 0) {
+        return Iterator.empty
+      }
+
+      val toDegree = getDegree(toNode, relTypes, dir.reversed, query)
+      if (toDegree == 0) {
```

Comment:
It can be zero for another relationship-type or direction. the isDense is a general thing.